ml-agents/mlagents/trainers/buffer.py

```diff
@@ -217,6 +220,48 @@ def make_mini_batch(self, start, end):
                 mini_batch[key] = np.array(self[key][start:end])
             return mini_batch
 
+        # SAC HAC
+        def sample_mini_batch(self, batch_size):
+            """
+            Creates a mini-batch from a random start and end.
+            : param batch_size: number of elements to withdraw.
+            """
+            mini_batch_lists = {}
+            mini_batch = Buffer.AgentBuffer()
+            idxes = [
+                random.randint(0, len(self["actions"]) - 1) for _ in range(batch_size)
+            ]
+
+            for i in idxes:
+                for key in self:
+                    if key in mini_batch_lists:
+                        mini_batch_lists[key].append(self[key][i])
+                    else:
+                        mini_batch_lists[key] = [self[key][i]]
+            for key in mini_batch_lists:
+                mini_batch[key] = np.array(mini_batch_lists[key])
+            return mini_batch
+
+        def save_to_file(self, filename):
```

Comment:
Might be more general here (and load) if you take a file-like object instead of requiring the filename: http://docs.h5py.org/en/stable/high/file.html#python-file-like-objects

Reply:
One downside to this is that this File must now be open for r/w, even when loading. I suppose this isn't an issue with our implementation as it's only loaded/saved once, so I moved to this. 

Might be useful for the distributed case.